pkg/virt-controller/watch/vmi.go

```diff
@@ -389,7 +393,7 @@ func (c *VMIController) hasOwnerVM(vmi *virtv1.VirtualMachineInstance) bool {
 	return false
 }
 
-func (c *VMIController) updateStatus(vmi *virtv1.VirtualMachineInstance, pod *k8sv1.Pod, dataVolumes []*cdiv1.DataVolume, syncErr syncError) error {
+func (c *VMIController) updateStatus(vmi *virtv1.VirtualMachineInstance, pod *k8sv1.Pod, dataVolumes []*cdiv1.DataVolume, syncErr syncError) (*virtv1.VirtualMachineInstance, syncError, error) {
```

Comment:
Why do we need to return 2 types of errors ?.... Can't we just live with `syncError`, as was done before ?

Reply:
It is a mess, I am just exposing it :)

The update of the object requires this. If you have an idea how we can avoid it, it will be great to get rid of it.

Reply:
Looking at it, maybe we can move:
```go
	reason := ""
	if syncErr != nil {
		reason = syncErr.Reason()
	}
	conditionManager.CheckFailure(vmiCopy, syncErr, reason)
```
Out of the function.

However, I am not sure how it affects the patch before it.

Reply:
You could move this block to the `updateStatus()` method, because it updates conditions. That function has a `switch` block checking VMI phase, so you could add this block only to cases that does not satisfy the condition above: `if vmi.IsRunning() || vmi.IsScheduled()`. Then the behavior would not be changed.

Reply:
I tried to do a syntactic refactoring, exposing clearly where the update is done.
I must admit, the code is unreadable to me and I am not sure if it can be changed without risking a bug.

For example, the first part argues that if the VMI is "owned" by the handler, it should just patch it and finish.
But there are multiple problems with this note:
- The VMI is owner is the virt-controller, no one else. The virt-handler should never call `Update` (but it does).
- The patching is limited to only a few sections: VolumeStatus, Conditions, ActivePods, LauncherContainerImageVersion and Labels.
  - On the other hand, the "update" part covers all `Status`, Finalizers, Annotations and Labels.
  - It also does other mutations, by calliing this CheckFailure and SetVMIPhaseTransitionTimestamp.

Any change here is a risk.

Reply:
Ok, I understand.

Reply:
it looks like we still have more code to improve